Implement legacy osu! cursor particles

Opelkuh opened the pull request:
Part of https://github.com/ppy/osu/issues/13760


https://user-images.githubusercontent.com/25430283/133524029-6b36c4ec-2026-4fa4-bba7-447301b089e4.mp4


First of all, I would like to ask for help with optimizing the "currently tracking Kiai hit object" lookup :eyes: While it works fine now, it's really not ideal. I'll let this image speak for itself:

![rider64](https://user-images.githubusercontent.com/25430283/133524160-6fcb8fff-0c0e-4d82-bed5-9d6f533e61d8.png)

I was thinking of exposing something similar to `Playfield.HitObjectUsageBegan` but with `DrawableHitObject`, but I decided that I would rather ask here than implement something hacky. Is there a better way to detect sliders/spinners that are currently clicked by the player?

There is also a visual difference between this PR and stable, which is that the particles stop earlier during the initial break before the first hit object appears. This means that they don't show up at the start at all on maps that don't have any initial break, while on stable they're visible on every map. This could be fixed by lowering `GameplayStartTime`, but I wasn't sure if that's feasible so I left it as is for now.

### Remarks
I made `ParticleSpewer` as an abstract class because I would like to reuse it for other parts of the original issue.

Opelkuh committed:
Add `ParticleJet`

Opelkuh committed:
Change `ParticleSpewer` to use screen space

Opelkuh committed:
Fix `ParticleSpewer` gravity calculation

Opelkuh committed:
Change `ParticleSpewer.Active` to a Bindable

Opelkuh committed:
Add legacy cursor star particles

Opelkuh committed:
Add cursor velocity to star particles

Opelkuh committed:
Add `ParticleParent` option to `ParticleSpewer`

Opelkuh committed:
Remove `StartScale` from `ParticleSpewer` particles

Opelkuh committed:
Change particle `AngularVelocity` into `StartAngle` and `EndAngle`

Opelkuh committed:
Move cursor particles under `OsuCursorContainer`

Opelkuh committed:
Remove `ParticleJet`

Opelkuh committed:
Add `OsuSkinComponents.CursorParticles`

Opelkuh committed:
Rename legacy cursor particle classes

Opelkuh committed:
Scale down cursor particles

Opelkuh committed:
Fix `LegacyCursorParticles` texture null reference

Opelkuh committed:
Add cursor particles tests

Opelkuh committed:
Fix outdated test step description

Opelkuh committed:
Remove Particles namespace

Opelkuh committed:
Fix code inspect failure

Opelkuh committed:
Tiny refactor

Opelkuh committed:
Add `StarBreakAdditive` config support

Opelkuh committed:
Remove hardcoded particle scale

Opelkuh committed:
Fix cursor particle scale

peppy committed:
Remove implied defaults

peppy committed:
`SpawnParticle` -> `CreateParticle` (and set time outside of `virtual` call)

Allows easier overriding (no need to call the `base.CreateParticle` call
and worry about overwriting the time value.

peppy committed:
Use `private` for internally used property

peppy commented:
Pretty nice work here. I've committed some menial fixes, but have a few points to bring up which should *probably* be addressed:

- I don't think the class needs to be abstract. It can probably be used as-is in a meaningful way. I guess the only thing stopping that is the fact that a lot of the settings are `virtual` rather than public settable. These days we tend to prefer public-set for such variables, but this change doesn't necessarily need to be made, just something I thought I'd point out.
- The fact that this extends from `Sprite` makes sizing a bit weird. The spewer is spewing in a larger area to the size of the drawable:

![20210916 170909 (dotnet)](https://user-images.githubusercontent.com/191335/133575280-2f323b97-cc1e-4b61-96e6-fd32567c602d.png)

We've done similar things in a few other places, but it would be idea if this could be resolved. I haven't looked into how easy this is to do, though.

- Building on the previous point, right now the `StartPosition` isn't in the local space of the drawable. I actually don't know what space system it's in. Works fine for the gameplay usages, but it would be idea if the following setup would create particles at the centre of the spawning area, and allow the RNG movement to use the `RelativePositionAxes` specification:

```diff
diff --git a/osu.Game.Tests/Visual/Gameplay/TestSceneParticleSpewer.cs b/osu.Game.Tests/Visual/Gameplay/TestSceneParticleSpewer.cs
index c259718a7a..2a07bddef0 100644
--- a/osu.Game.Tests/Visual/Gameplay/TestSceneParticleSpewer.cs
+++ b/osu.Game.Tests/Visual/Gameplay/TestSceneParticleSpewer.cs
@@ -30,6 +30,7 @@ private void load()
             AddToggleStep("toggle spawning", value => spewer.Active.Value = value);
             AddSliderStep("particle gravity", 0f, 250f, 0f, value => spewer.Gravity = value);
             AddSliderStep("particle velocity", 0f, 500f, 250f, value => spewer.MaxVelocity = value);
+            AddStep("move to new location", () => spewer.TransformTo(nameof(spewer.SpawnPosition), new Vector2(RNG.NextSingle(), RNG.NextSingle()), 1000, Easing.Out));
         }
 
         [SetUpSteps]
@@ -51,14 +52,11 @@ public void TestPresence()
             AddAssert("is not present", () => !spewer.IsPresent);
         }
 
-        private TestParticleSpewer createSpewer()
-        {
-            return new TestParticleSpewer(skinManager.DefaultLegacySkin.GetTexture("star2"))
+        private TestParticleSpewer createSpewer() =>
+            new TestParticleSpewer(skinManager.DefaultLegacySkin.GetTexture("star2"))
             {
-                Anchor = Anchor.Centre,
-                Origin = Anchor.Centre,
+                RelativePositionAxes = Axes.Both,
             };
-        }
 
         private class TestParticleSpewer : ParticleSpewer
         {
@@ -66,8 +64,11 @@ private class TestParticleSpewer : ParticleSpewer
             private const int rate = 250;
 
             public float Gravity;
+
             public float MaxVelocity = 250;
 
+            public Vector2 SpawnPosition { get; set; } = new Vector2(0.5f);
+
             protected override float ParticleGravity => Gravity;
 
             public TestParticleSpewer(Texture texture)
@@ -82,6 +83,7 @@ public TestParticleSpewer(Texture texture)
                         RNG.NextSingle(-MaxVelocity, MaxVelocity),
                         RNG.NextSingle(-MaxVelocity, MaxVelocity)
                     ),
+                    StartPosition = SpawnPosition,
                     Duration = RNG.NextSingle(lifetime),
                     StartAngle = RNG.NextSingle(MathF.PI * 2),
                     EndAngle = RNG.NextSingle(MathF.PI * 2),
```

peppy reviewed (changes requested):
As proposed

Opelkuh committed:
Add `RelativePositionAxes` support

Opelkuh committed:
Normalize particle velocity based on max duration

Opelkuh committed:
Remove abstract from `ParticleSpewer`

Opelkuh commented:
Unless I'm missing something, `StartPosition` should've been in the local space as at the end it's being transformed with the DrawNode's draw matrix. When I tested your patch it was spawning in the top left corner for me.

I changed `ParticleSpewer` to not be abstract as suggested, but imo it's less readable now, primarily having the particle spawning in an outside delegate. I also presume that having `ParticleSpewer.FallingParticle` everywhere isn't ideal, would it ok if I split it out and move everything into a new `osu.Game.Graphics.Particles` namespace? (along with the `ParticleExplosion`)

I also added `RelativePositionAxes` support and changed the velocity calculation so that it's consistent when you change the maximum particle duration. I'm not sure how to completely deal with the sizing issue though, even if you set particle gravity to 0.5, it will pull particles that start with downwards velocity out of the spewer area:

![image](https://user-images.githubusercontent.com/25430283/133909630-badc621d-24ef-46db-8345-18d3ce2f09de.png)

Same goes with changing the spawn position to something that's not the center, you would have to adjust the velocity so that it compensates for how close the spawn position is to the edge. This could be done by the `ParticleSpewer` itself but imho that would make it a lot harder to tune the velocity for different visual effects.
`LegacyCursorParticles` would also have to be scaled to the whole screen for it to have the correct size, which would make it larger than the cursor container that it's in. For now I set the size to be 1:1 with the cursor container, same as `CursorTrail` that's under the same container.

peppy commented:
having things fall out of the area due to gravity is fine, i was only taking issue with the spawning point specifically. it sounds like you may have resolved the issue, but in my testing it was specifically relative position that wasn't working.

if you think it reads better abstract it's fine to leave it as such, not too fussed on that until we find more usages of this.

Opelkuh committed:
Merge branch 'master' into add-legacy-star-particles

Opelkuh committed:
Fix build errors

bdach commented on osu.Game/Graphics/ParticleSpewer.cs:
```diff
@@ -0,0 +1,208 @@
+            private Axes relativePositionAxes;
+            private Vector2 sourceSize;
+
+            public ParticleSpewerDrawNode(Sprite source)
```
Not sure if this is just me but with the cast in `Source` (the property) I'd prefer that this ctor parameter was scoped down to `ParticleSpewer`. I was really confused for a second as to what a sprite was doing here.

bdach commented on osu.Game/Graphics/ParticleSpewer.cs:
```diff
@@ -0,0 +1,208 @@
+            {
+                foreach (var p in particles)
+                {
+                    // ignore particles that weren't initialized.
+                    if (p.StartTime <= 0) continue;
```
This isn't 100% right. The start time can be earlier than 0 and still not mean that the particle wasn't initialised. That will be the case with beatmaps with a long lead-in time / long storyboarded intro, and in those cases the gameplay clock does not start at time 0.

An example map that this can be seen on is https://osu.ppy.sh/beatmapsets/5381#osu/26567 (stable still shows the particles even in the intro section).

Opelkuh replied:
Should be fixed now. I also added a test for these cases.

bdach commented on osu.Game/Graphics/ParticleSpewer.cs:
```diff
@@ -0,0 +1,208 @@
+
+            public float AlphaAtTime(float timeSinceStart) => 1 - progressAtTime(timeSinceStart);
+
+            public float ScaleAtTime(float timeSinceStart) => 1 + (EndScale - 1) * progressAtTime(timeSinceStart);
+
+            public float AngleAtTime(float timeSinceStart) => StartAngle + (EndAngle - StartAngle) * progressAtTime(timeSinceStart);
```
Opportunity to use `Interpolation.Lerp` here (unless inlining like that is faster? but not sure it would be)

Opelkuh replied:
`Interpolation.Lerp` only works with doubles so there has to be a cast to float, but that's most likely a negligible performance difference so I went with it.

bdach commented on osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs:
```diff
@@ -0,0 +1,205 @@
+            if (osuPlayfield == null) return;
+
+            // find active kiai slider or spinner.
+            var kiaiHitObject = osuPlayfield.HitObjectContainer.AliveObjects.FirstOrDefault(h =>
+                h.HitObject.Kiai &&
+                (
+                    (h is DrawableSlider slider && slider.Tracking.Value) ||
+                    (h is DrawableSpinner spinner && spinner.RotationTracker.Tracking)
+                )
+            );
```
When it comes to any allocs here, they will be due to underlying LINQ iterator usage. I don't necessarily thing anything smart can be done regarding them in this pull, it should probably be fixed holistically separately (because other places also use `AliveObjects` in similar ways).

If `OrderedDictionary` wasn't such a terrible collection type it could have been used in `PooledDrawableWithLifetimeContainer`, but alas...

Opelkuh replied:
Yeah it might be easiest to leave it as is for now, I tested it further and the performance hit of this seems to be smaller than I originally thought (at least on my desktop PC).

I think having something similar to `ApplyToDrawableHitObject` that mods use would be ideal as I'd only need to bind the `Tracking` bindable when a hitobject enters the playfield.

peppy replied:
I've pushed a solution to avoid this enumeration outside of kiai mode, which I think is an improvement (even though `AliveObjects` is used in other places, it *is* expensive and should be avoided for always-on cases like this where possible).

@bdach Can you take a look and make sure you're okay with the change?

bdach replied:
Yep, all good with me.

Opelkuh committed:
Revert `af4c3727d77a16e2534df9bbf452336b5c544342`

Opelkuh committed:
Fix particles not spawning if `Time.Current` is negative

Opelkuh committed:
Scope down DrawNode's `source` parameter

Opelkuh committed:
Use `Interpolation.Lerp`

Opelkuh committed:
Add test for time jumps

Opelkuh commented:
I reverted the abstract change for now. Sorry @bdach for doing it mid-review, I noticed your comments too late.

peppy committed:
Merge branch 'master' into add-legacy-star-particles

peppy commented on .idea/.idea.osu/.idea/indexLayout.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ContentModelUserStore">
+  <component name="UserContentModel">
```
This suggests you're not using the `osu.Desktop.slnf` project, which you probably should be (it doesn't contain mobile projects and therefore is quite a lot faster to work with).

Probably okay to just let this one pass as I believe this is the new correct value, just hasn't been applied yet because no one uses the main `sln` file.

Opelkuh replied:
Ah sorry, I didn't mean to commit that. You're right, I was using `osu.sln` but had all the non-desktop stuff unloaded. Rider also automatically changed `.idea/.idea.osu/.idea/projectSettingsUpdater.xml` and deleted `.idea/.idea.osu/.idea/modules.xml`.

I'll use `osu.Desktop.slnf` in the future.

peppy committed:
Avoid `AliveObject` enumeration when not in kiai section
